diff -ur qt-x11-opensource-src-4.5.0-o/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.cpp qt-x11-opensource-src-4.5.0/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.cpp
--- qt-x11-opensource-src-4.5.0-o/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.cpp	2009-04-05 12:42:47.000000000 +0200
+++ qt-x11-opensource-src-4.5.0/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.cpp	2009-04-05 12:59:36.000000000 +0200
@@ -52,6 +52,7 @@
     QHash<int, int> fontSizes;
     QHash<int, bool> attributes;
     QUrl userStyleSheetLocation;
+    QString defaultTextEncoding;
     QString localStorageDatabasePath;
     QString offlineWebApplicationCachePath;
     qint64 offlineStorageDefaultQuota;
@@ -161,6 +162,9 @@
         QUrl location = !userStyleSheetLocation.isEmpty() ? userStyleSheetLocation : global->userStyleSheetLocation;
         settings->setUserStyleSheetLocation(WebCore::KURL(location));
 
+        QString encoding = !defaultTextEncoding.isEmpty() ? defaultTextEncoding: global->defaultTextEncoding;
+        settings->setDefaultTextEncodingName(encoding);
+
         QString localStoragePath = !localStorageDatabasePath.isEmpty() ? localStorageDatabasePath : global->localStorageDatabasePath;
         settings->setLocalStorageDatabasePath(localStoragePath);
 
@@ -425,6 +429,29 @@
 }
 
 /*!
+    Specifies the default encoding system
+
+    The \a encoding, must be a string describing an encoding.
+
+    \sa userStyleSheetUrl()
+*/
+void QWebSettings::setDefaultTextEncoding(const QString &encoding) 
+{
+    d->defaultTextEncoding = encoding;
+    d->apply();
+}
+
+/*!
+    Returns the default text encoding
+
+    \sa setDefaultTextEncoding()
+*/
+QString QWebSettings::defaultTextEncoding() const
+{
+    return d->defaultTextEncoding;
+}
+
+/*!
     Sets the path of the icon database to \a path. The icon database is used
     to store "favicons" associated with web sites.
 
diff -ur qt-x11-opensource-src-4.5.0-o/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.h qt-x11-opensource-src-4.5.0/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.h
--- qt-x11-opensource-src-4.5.0-o/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.h	2009-04-05 12:42:47.000000000 +0200
+++ qt-x11-opensource-src-4.5.0/src/3rdparty/webkit/WebKit/qt/Api/qwebsettings.h	2009-04-05 12:46:45.000000000 +0200
@@ -95,6 +95,9 @@
     void setUserStyleSheetUrl(const QUrl &location);
     QUrl userStyleSheetUrl() const;
 
+    void setDefaultTextEncoding(const QString &encoding);
+    QString defaultTextEncoding() const;
+
     static void setIconDatabasePath(const QString &location);
     static QString iconDatabasePath();
     static void clearIconDatabase();
