--- qt-x11-opensource-src-4.5.1.orig/src/3rdparty/webkit/WebCore/page/PrintContext.h	2009-04-22 01:57:40.000000000 +0200
+++ qt-x11-opensource-src-4.5.1/src/3rdparty/webkit/WebCore/page/PrintContext.h	2009-06-20 12:21:48.630007375 +0200
@@ -50,6 +50,8 @@
 protected:
     Frame* m_frame;
     Vector<IntRect> m_pageRects;
+public:
+    const Vector<IntRect> & getPageRects() {return m_pageRects;}
 };
 
 }

diff -ur qt-x11-opensource-src-4.5.0-old/src/3rdparty/webkit/WebCore/page/PrintContext.h qt-x11-opensource-src-4.5.0/src/3rdparty/webkit/WebCore/page/PrintContext.h
