diff -ur qt-x11-opensource-src-4.5.1.orig/src/gui/kernel/qapplication_x11.cpp qt-x11-opensource-src-4.5.1/src/gui/kernel/qapplication_x11.cpp
--- qt-x11-opensource-src-4.5.1.orig/src/gui/kernel/qapplication_x11.cpp	2009-04-22 01:57:41.000000000 +0200
+++ qt-x11-opensource-src-4.5.1/src/gui/kernel/qapplication_x11.cpp	2009-06-20 14:58:44.986333611 +0200
@@ -2215,6 +2217,8 @@
         QSegfaultHandler::initialize(priv->argv, priv->argc);
 #endif
         QCursorData::initialize();
+    } else if (!QApplicationPrivate::graphics_system_name.isNull()) {
+        QApplicationPrivate::graphics_system = QGraphicsSystemFactory::create(QApplicationPrivate::graphics_system_name);
     }
     QFont::initialize();
 
diff -ur qt-x11-opensource-src-4.5.1.orig/src/gui/styles/qstyle.cpp qt-x11-opensource-src-4.5.1/src/gui/styles/qstyle.cpp
--- qt-x11-opensource-src-4.5.1.orig/src/gui/styles/qstyle.cpp	2009-04-22 01:57:41.000000000 +0200
+++ qt-x11-opensource-src-4.5.1/src/gui/styles/qstyle.cpp	2009-06-20 12:00:14.946507441 +0200
@@ -47,6 +47,7 @@
 #include "qpixmapcache.h"
 #include "qstyleoption.h"
 #include "private/qstyle_p.h"
+#include "private/qapplication_p.h"
 #ifndef QT_NO_DEBUG
 #include "qdebug.h"
 #endif
@@ -2211,7 +2212,7 @@
 {
 #ifdef Q_WS_X11
     QColor background;
-    if (QX11Info::appDepth() > 8)
+    if (!qt_is_gui_used || QX11Info::appDepth() > 8)
         background = QColor(0xd4, 0xd0, 0xc8); // win 2000 grey
     else
         background = QColor(192, 192, 192);
diff -ur qt-x11-opensource-src-4.5.1.orig/src/gui/image/qpixmap_raster.cpp qt-x11-opensource-src-4.5.1/src/gui/image/qpixmap_raster.cpp
--- qt-x11-opensource-src-4.5.1.orig/src/gui/image/qpixmap_raster.cpp	2009-04-22 01:57:41.000000000 +0200
+++ qt-x11-opensource-src-4.5.1/src/gui/image/qpixmap_raster.cpp	2009-06-20 12:00:14.934006602 +0200
@@ -44,6 +44,7 @@
 #include "qpixmap_raster_p.h"
 #include "qnativeimage_p.h"
 #include "qimage_p.h"
+#include "kernel/qapplication_p.h"
 
 #include "qbitmap.h"
 #include "qimage.h"
@@ -89,8 +90,10 @@
 #else
     if (pixelType() == BitmapType)
         format = QImage::Format_MonoLSB;
-    else
+    else if (qt_is_gui_used)
         format = QNativeImage::systemFormat();
+    else
+        format = QImage::Format_RGB32;
 #endif
 
     image = QImage(width, height, format);
@@ -156,7 +159,7 @@
                     : sourceImage.convertToFormat(QImage::Format_RGB32);
         } else {
 
-            QImage::Format opaqueFormat = QNativeImage::systemFormat();
+            QImage::Format opaqueFormat = qt_is_gui_used ? QNativeImage::systemFormat() : QImage::Format_RGB32;
             QImage::Format alphaFormat = QImage::Format_ARGB32_Premultiplied;
 
             switch (opaqueFormat) {
